In [2]:
import pandas as pd
import numpy as np
train = pd.read_csv('./train.csv')

In [3]:
df = train.sample(n=100)

df.head(10)

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
1512           0.275966          -0.016343          -0.106286   
4793           0.296508          -0.015842          -0.108012   
4364           0.260883          -0.044814          -0.105528   
7277           0.237512           0.004841          -0.062039   
3447           0.268900          -0.026059          -0.135966   
819            0.249910          -0.003232          -0.132582   
3808           0.279801          -0.028728          -0.121241   
6661           0.275539          -0.018613          -0.106569   
3425           0.207159           0.038422          -0.106634   
1559           0.214767          -0.018180          -0.102710   

      tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  \
1512         -0.997062         -0.991467         -0.959090         -0.997297   
4793         -0.960069         -0.981255         -0.980927         -0.959123   
4364          0.235884          0.342620         -0.059147          0.210284   
7277         -0.360827         -0.073761         -0.118117         -0.432977   
3447         -0.995742         -0.986258         -0.962409         -0.996276   
819           0.411690          0.381539         -0.235722          0.389632   
3808         -0.986804         -0.938825         -0.944882         -0.988213   
6661         -0.997265         -0.991787         -0.995409         -0.997387   
3425         -0.085772         -0.185069         -0.085050         -0.218795   
1559         -0.253919          0.078742          0.052856         -0.276134   

      tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  tBodyAcc-max()-X  \
1512         -0.992030         -0.955360         -0.942048   
4793         -0.977839         -0.979861         -0.902849   
4364          0.320321         -0.084299          0.326680   
7277         -0.107902         -0.105311         -0.159473   
3447         -0.985498         -0.957563         -0.940637   
819           0.181010         -0.250503          0.595843   
3808         -0.940756         -0.949459         -0.932228   
6661         -0.992188         -0.995786         -0.944131   
3425         -0.211111         -0.101418          0.351485   
1559          0.016020          0.058896         -0.028944   

             ...          fBodyBodyGyroJerkMag-kurtosis()  \
1512         ...                                -0.489543   
4793         ...                                -0.399175   
4364         ...                                -0.895754   
7277         ...                                 0.108854   
3447         ...                                -0.873167   
819          ...                                -0.637409   
3808         ...                                -0.836289   
6661         ...                                -0.950764   
3425         ...                                -0.946281   
1559         ...                                -0.190089   

      angle(tBodyAccMean,gravity)  angle(tBodyAccJerkMean),gravityMean)  \
1512                     0.228398                             -0.291504   
4793                    -0.108392                             -0.385004   
4364                    -0.047803                              0.785543   
7277                     0.304911                              0.181551   
3447                    -0.014510                              0.644098   
819                      0.322894                              0.813965   
3808                    -0.052661                              0.714628   
6661                     0.206001                              0.408623   
3425                     0.395526                              0.499935   
1559                     0.853708                              0.377824   

      angle(tBodyGyroMean,gravityMean)  angle(tBodyGyroJerkMean,gravityMean)  \
1512                         -0.122568                              0.498036   
4793                          0.034504                              

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 100 entries, 1512 to 5826
Columns: 563 entries, tBodyAcc-mean()-X to Activity
dtypes: float64(561), int64(1), object(1)
memory usage: 440.6+ KB


In [5]:
df.describe()

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
count         100.000000         100.000000         100.000000   
mean            0.287385          -0.018529          -0.110429   
std             0.060896           0.026096           0.037013   
min             0.128784          -0.133438          -0.267197   
25%             0.264187          -0.026960          -0.121410   
50%             0.278300          -0.017046          -0.108686   
75%             0.295566          -0.013229          -0.100437   
max             0.459935           0.094467           0.022727   

       tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  \
count        100.000000        100.000000        100.000000        100.000000   
mean          -0.534404         -0.457194         -0.575372         -0.563273   
std            0.493869          0.519339          0.409123          0.468607   
min           -0.997884         -0.995383         -0.997241         -0.998081   
25%           -0.993458         -0.979008         -0.976339         -0.994389   
50%           -0.755707         -0.374777         -0.679153         -0.779769   
75%           -0.136688          0.027902         -0.204321         -0.190087   
max            0.469275          0.447302          0.253409          0.463658   

       tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  tBodyAcc-max()-X     ...      \
count        100.000000        100.000000        100.000000     ...       
mean          -0.478075         -0.570103         -0.378491     ...       
std            0.498407          0.414215          0.589572     ...       
min           -0.993890         -0.997647         -0.944131     ...       
25%           -0.978352         -0.975245         -0.935393     ...       
50%           -0.400621         -0.678710         -0.602336     ...       
75%           -0.016138         -0.212730          0.138185     ...       
max            0.371491          0.280505          0.624140     ...       

       fBodyBodyGyroJerkMag-skewness()  fBodyBodyGyroJerkMag-kurtosis()  \
count                       100.000000                       100.000000   
mean                         -0.324235                        -0.649044   
std                           0.285998                         0.275610   
min                          -0.859261                        -0.966703   
25%                          -0.539759                        -0.831496   
50%                          -0.380590                        -0.732573   
75%                          -0.146072                        -0.541110   
max                           0.579021                         0.423149   

       angle(tBodyAccMean,gravity)  angle(tBodyAccJerkMean),gravityMean)  \
count                   100.000000                            100.000000   
mean                     -0.038360                              0.028051   
std                       0.367309                              0.476903   
min                      -0.850387                             -1.000000   
25%                      -0.182341                             -0.287171   
50%                      -0.010702                              0.060873   
75%                       0.154948                              0.347691   
max                       0.853708                              0.951194   

       angle(tBodyGyroMean,gravityMean)  angle(tBodyGyroJerkMean,gravityMean)  \
count                        100.000000                            100.000000   
mean                          -0.065230                             -0.020497   
std                            0.658317                              0.462019   
min                           -0.981914                             -0.938698   
25%                           -0.742141                             -0.436561   
50%                           -0.086798                              0.056338   
75%                            0.473356                              0.322281 

In [6]:
#Creating the dependent variable class
factor = pd.factorize(df['Activity'])
df.Activity = factor[0]
definitions = factor[1]
print(df.Activity.head())
print(definitions)


1512    0
4793    1
4364    2
7277    3
3447    4
Name: Activity, dtype: int64
Index(['SITTING', 'LAYING', 'WALKING_DOWNSTAIRS', 'WALKING', 'STANDING',
       'WALKING_UPSTAIRS'],
      dtype='object')


In [7]:

#Splitting the data into independent and dependent variables
X = df.iloc[:,0:-1].values
y = df.iloc[:,-1].values
print('The independent features set: ')
print(X[:5,:])
print('The dependent variable: ')
print(y[:5])

The independent features set: 
[[ 2.7596619e-01 -1.6342746e-02 -1.0628628e-01 ...  9.9144146e-02
  -2.8899787e-02  7.0000000e+00]
 [ 2.9650818e-01 -1.5841751e-02 -1.0801154e-01 ... -2.5567828e-01
  -7.5631624e-01  2.3000000e+01]
 [ 2.6088341e-01 -4.4814456e-02 -1.0552814e-01 ...  3.2357039e-01
   1.4482115e-01  2.1000000e+01]
 [ 2.3751176e-01  4.8407904e-03 -6.2038682e-02 ...  1.8430866e-01
   4.7131898e-03  3.0000000e+01]
 [ 2.6889989e-01 -2.6059057e-02 -1.3596622e-01 ...  3.1370389e-01
   6.6395383e-02  1.7000000e+01]]
The dependent variable: 
[0 1 2 3 4]


In [8]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 21)

In [9]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [10]:
from sklearn.ensemble import RandomForestClassifier

classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 42)
classifier.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='entropy',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=None,
            oob_score=False, random_state=42, verbose=0, warm_start=False)

In [11]:
# Predicting the Test set results
y_pred = classifier.predict(X_test)
#Reverse factorize (converting y_pred from 0s,1s and 2s to Iris-setosa, Iris-versicolor and Iris-virginica
reversefactor = dict(zip(range(6),definitions))
y_test = np.vectorize(reversefactor.get)(y_test)
y_pred = np.vectorize(reversefactor.get)(y_pred)
# Making the Confusion Matrix
print(pd.crosstab(y_test, y_pred, rownames=['Actual Species'], colnames=['Predicted Species']))

Predicted Species   LAYING  SITTING  STANDING  WALKING  WALKING_DOWNSTAIRS  \
Actual Species                                                               
LAYING                   4        0         0        0                   0   
SITTING                  0        2         0        0                   0   
STANDING                 0        1         4        0                   0   
WALKING                  0        0         0        7                   1   
WALKING_DOWNSTAIRS       0        0         0        0                   2   
WALKING_UPSTAIRS         0        0         0        0                   1   

Predicted Species   WALKING_UPSTAIRS  
Actual Species                        
LAYING                             0  
SITTING                            0  
STANDING                           0  
WALKING                            0  
WALKING_DOWNSTAIRS                 0  
WALKING_UPSTAIRS                   3  


In [12]:
from sklearn.model_selection import cross_val_score

scores = cross_val_score(classifier, X, y, cv=5)
print(scores)

[1.         0.85714286 0.9        0.94736842 0.66666667]


In [13]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.multiclass import OneVsRestClassifier
from sklearn.neural_network import MLPClassifier

In [26]:
def print_accuracy_report(classifier, X, y, num_validations=5):
    accuracy = cross_val_score(classifier, 
            X, y, scoring='accuracy', cv=num_validations)
    print("Accuracy: " + str(round(100*accuracy.mean(), 2)) + "%")

    f1 = cross_val_score(classifier, 
            X, y, scoring='f1_weighted', cv=num_validations)
    print("F1: " + str(round(100*f1.mean(), 2)) + "%")

    precision = cross_val_score(classifier, 
            X, y, scoring='precision_weighted', cv=num_validations)
    print("Precision: " + str(round(100*precision.mean(), 2)) + "%")

    recall = cross_val_score(classifier, 
            X, y, scoring='recall_weighted', cv=num_validations)
    print("Recall: " + str(round(100*recall.mean(), 2)) + "%" )

def trainResult(df, classifier):
    factor = pd.factorize(df['Activity'])
    df.Activity = factor[0]
    definitions = factor[1]

    X = df.iloc[:,0:-1].values
    y = df.iloc[:,-1].values

#     print(X, y)
    
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 21)

    scaler = StandardScaler()
    X_train = scaler.fit_transform(X_train)
    X_test = scaler.transform(X_test)

#     classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 42)
    classifier.fit(X_train, y_train)

    y_pred = classifier.predict(X_test)
    reversefactor = dict(zip(range(6),definitions))
    y_test = np.vectorize(reversefactor.get)(y_test)
    y_pred = np.vectorize(reversefactor.get)(y_pred)

    # Making the Confusion Matrix
#     print(pd.crosstab(y_test, y_pred, rownames=['Actual Species'], colnames=['Predicted Species']))

    scores = print_accuracy_report(classifier, X, y)
    print()
    return

In [28]:
classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 42)
print('n = 100, RandomForestClassifier')
trainResult(train.sample(n=100), classifier)
print('n = 500, RandomForestClassifier')
trainResult(train.sample(n=500), classifier)
print('n = 1000, RandomForestClassifier')
trainResult(train.sample(n=1000), classifier)
print('n = 3000, RandomForestClassifier')
trainResult(train.sample(n=3000), classifier)
print('n = 5000, RandomForestClassifier')
trainResult(train.sample(n=5000), classifier)
print('n = all, RandomForestClassifier')
trainResult(train, classifier)

n = 100, RandomForestClassifier
Accuracy: 77.97%
F1: 77.83%
Precision: 83.24%
Recall: 77.97%

n = 500, RandomForestClassifier
Accuracy: 91.02%
F1: 90.96%
Precision: 91.81%
Recall: 91.02%

n = 1000, RandomForestClassifier
Accuracy: 91.89%
F1: 91.84%
Precision: 92.28%
Recall: 91.89%

n = 3000, RandomForestClassifier
Accuracy: 95.83%
F1: 95.84%
Precision: 95.89%
Recall: 95.83%

n = 5000, RandomForestClassifier
Accuracy: 96.58%
F1: 96.58%
Precision: 96.59%
Recall: 96.58%

n = all, RandomForestClassifier
Accuracy: 90.77%
F1: 90.66%
Precision: 91.2%
Recall: 90.77%



In [29]:
classifier = OneVsRestClassifier(KNeighborsClassifier(),-1)
print('n = 100, KNeighborsClassifier')
trainResult(train.sample(n=100), classifier)
print('n = 500, KNeighborsClassifier')
trainResult(train.sample(n=500), classifier)
print('n = 1000, KNeighborsClassifier')
trainResult(train.sample(n=1000), classifier)
print('n = 3000, KNeighborsClassifier')
trainResult(train.sample(n=3000), classifier)
print('n = 5000, KNeighborsClassifier')
trainResult(train.sample(n=5000), classifier)
print('n = all, KNeighborsClassifier')
trainResult(train, classifier)

n = 100, KNeighborsClassifier
Accuracy: 58.91%
F1: 55.28%
Precision: 56.84%
Recall: 58.91%

n = 500, KNeighborsClassifier
Accuracy: 80.59%
F1: 80.03%
Precision: 81.47%
Recall: 80.59%

n = 1000, KNeighborsClassifier
Accuracy: 88.4%
F1: 88.29%
Precision: 88.92%
Recall: 88.4%

n = 3000, KNeighborsClassifier
Accuracy: 94.3%
F1: 94.29%
Precision: 94.4%
Recall: 94.3%

n = 5000, KNeighborsClassifier
Accuracy: 95.64%
F1: 95.62%
Precision: 95.72%
Recall: 95.64%

n = all, KNeighborsClassifier
Accuracy: 80.23%
F1: 79.81%
Precision: 82.71%
Recall: 80.23%



In [30]:
classifier = OneVsRestClassifier(LogisticRegression(),-1)
print('n = 100, LogisticRegression')
trainResult(train.sample(n=100), classifier)
print('n = 500, LogisticRegression')
trainResult(train.sample(n=500), classifier)
print('n = 1000, LogisticRegression')
trainResult(train.sample(n=1000), classifier)
print('n = 3000, LogisticRegression')
trainResult(train.sample(n=3000), classifier)
print('n = 5000, LogisticRegression')
trainResult(train.sample(n=5000), classifier)
print('n = all, LogisticRegression')
trainResult(train, classifier)

n = 100, LogisticRegression
Accuracy: 92.97%
F1: 91.9%
Precision: 92.28%
Recall: 92.97%

n = 500, LogisticRegression
Accuracy: 93.39%
F1: 93.39%
Precision: 93.66%
Recall: 93.39%

n = 1000, LogisticRegression
Accuracy: 96.02%
F1: 96.02%
Precision: 96.1%
Recall: 96.02%

n = 3000, LogisticRegression
Accuracy: 97.77%
F1: 97.76%
Precision: 97.81%
Recall: 97.77%

n = 5000, LogisticRegression
Accuracy: 97.96%
F1: 97.96%
Precision: 97.97%
Recall: 97.96%

n = all, LogisticRegression
Accuracy: 93.77%
F1: 93.74%
Precision: 94.4%
Recall: 93.77%



In [31]:
classifier = OneVsRestClassifier(MLPClassifier(),-1)
print('n = 100, MLPClassifier')
trainResult(train.sample(n=100), classifier)
print('n = 500, MLPClassifier')
trainResult(train.sample(n=500), classifier)
print('n = 1000, MLPClassifier')
trainResult(train.sample(n=1000), classifier)
print('n = 3000, MLPClassifier')
trainResult(train.sample(n=3000), classifier)
print('n = 5000, MLPClassifier')
trainResult(train.sample(n=5000), classifier)
print('n = all, MLPClassifier')
trainResult(train, classifier)

n = 100, MLPClassifier
Accuracy: 84.72%
F1: 85.41%
Precision: 90.2%
Recall: 82.98%

n = 500, MLPClassifier
Accuracy: 91.7%
F1: 91.81%
Precision: 93.68%
Recall: 91.31%

n = 1000, MLPClassifier
Accuracy: 94.48%
F1: 93.81%
Precision: 93.58%
Recall: 91.76%

n = 3000, MLPClassifier
Accuracy: 97.83%
F1: 97.67%
Precision: 96.86%
Recall: 97.53%

n = 5000, MLPClassifier
Accuracy: 98.36%
F1: 98.0%
Precision: 98.23%
Recall: 98.08%

n = all, MLPClassifier
Accuracy: 93.41%
F1: 93.17%
Precision: 94.08%
Recall: 92.9%



In [1]:
import warnings
warnings.simplefilter('ignore')


## Reference

* https://www.codementor.io/agarrahul01/multiclass-classification-using-random-forest-on-scikit-learn-library-hkk4lwawu
* https://www.programcreek.com/python/example/75177/sklearn.cross_validation.cross_val_score

(1) RandomForestClassifier
n = 100  => Accuracy: 77.97% F1: 77.83% Precision: 83.24% Recall: 77.97%
n = 500  => Accuracy: 91.02% F1: 90.96% Precision: 91.81% Recall: 91.02%
n = 1000 => Accuracy: 91.89% F1: 91.84% Precision: 92.28% Recall: 91.89%
n = 3000 => Accuracy: 95.83% F1: 95.84% Precision: 95.89% Recall: 95.83%
n = 5000 => Accuracy: 96.58% F1: 96.58% Precision: 96.59% Recall: 96.58%

(2) KNeighborsClassifier
n = 100  => Accuracy: 58.91% F1: 55.28% Precision: 56.84% Recall: 58.91%
n = 500  => Accuracy: 80.59% F1: 80.03% Precision: 81.47% Recall: 80.59%
n = 1000 => Accuracy: 88.4%  F1: 88.29% Precision: 88.92% Recall: 88.4%
n = 3000 => Accuracy: 94.3%  F1: 94.29% Precision: 94.4%  Recall: 94.3%
n = 5000 => Accuracy: 95.64% F1: 95.62% Precision: 95.72% Recall: 95.64%

(3) LogisticRegression
n = 100  => Accuracy: 92.97% F1: 91.9%  Precision: 92.28% Recall: 92.97%
n = 500  => Accuracy: 93.39% F1: 93.39% Precision: 93.66% Recall: 93.39%
n = 1000 => Accuracy: 96.02% F1: 96.02% Precision: 96.1%  Recall: 96.02%
n = 3000 => Accuracy: 97.77% F1: 97.76% Precision: 97.81% Recall: 97.77%
n = 5000 => Accuracy: 97.96% F1: 97.96% Precision: 97.97% Recall: 97.96%

(4) MLPClassifier
n = 100  => Accuracy: 84.72% F1: 85.41% Precision: 90.2%  Recall: 82.98%
n = 500  => Accuracy: 91.7%  F1: 91.81% Precision: 93.68% Recall: 91.31%
n = 1000 => Accuracy: 94.48% F1: 93.81% Precision: 93.58% Recall: 91.76%
n = 3000 => Accuracy: 97.83% F1: 97.67% Precision: 96.86% Recall: 97.53%
n = 5000 => Accuracy: 98.36% F1: 98.0%  Precision: 98.23% Recall: 98.08%
